### PR TITLE
Refactoring board and computer tests

### PR DESCRIPTION
**Pull request by fabientownsend**

### CODE REVIEW

File: lib/board.rb [hunk 1]
```diff
@@ -34,8 +34,48 @@ def free_positions
     board.flatten.select { |cell| cell.is_a?(Integer) }
   end
 
+  def string_to_board(board_string)
```
Comment (felipesere) [inline lib/board.rb]: How come this is an instance method? This means I need an instance of a board FIRST before I can fill it up with data from the string...

File: spec/board_spec.rb [hunk 2]
```diff
@@ -5,67 +5,75 @@
...
-
-  it "should be free position when it never been set" do
+  it "should be a free position when position free" do
     position = 0
+    board.string_to_board("   ,   ,   ")
     expect(board.free_positions.include?(position)).to be true
   end
 
-  it "should shouldn't get the spot after it set" do
+  it "shouldn't be a free position when postion isn't free" do
```
Comment (felipesere) [inline spec/board_spec.rb]: This is describg itself with itself...

File: lib/board.rb (hunk 1 above)
Comment (fabientownsend) [reply lib/board.rb]: I'm thinking about it, I have to consider that I will initialise the board with its size. May make it able to initialise the board with two different method.

Board.new(3)
Board.new("   ,    ,    ")

Ok spaces look weird

File: spec/board_spec.rb (hunk 2 above)
Comment (fabientownsend) [reply spec/board_spec.rb]: Recursion!

File: lib/board.rb (hunk 1 above)
Comment (christophgockel) [reply lib/board.rb]: +1, this method doesn't really belong here, does it?
Sometimes it's okay to have methods in the production code to aid testing or increase testability. In this case though it feels like a better place would be a helper method in the test file I think.

File: lib/board.rb [hunk 3]
```diff
@@ -34,8 +34,48 @@ def free_positions
...
+          board[index_row][index_column] = Mark::CROSS
+        elsif rows[index_row].chars[index_column] == Mark::ROUND
+          @counter += 1
+          board[index_row][index_column] = Mark::ROUND
+        end
+      end
+    end
+  end
+
+  def to_s
```
Comment (christophgockel) [inline lib/board.rb]: I advocate against using `to_s` for business specific behaviour.
It's preferred to have a specific method for that as `to_s` (similar to Java's `toString()` method) should not have domain specific knowledge. I always think of `to_s` and `toString()` as methods that might help me log a certain object, etc.

Additionally to that the behaviour this method implements is really a UI concern. Why should the `Board` itself know about the details about where to put spaces and commas?

File: spec/board_spec.rb (hunk 2 above)
Comment (christophgockel) [reply spec/board_spec.rb]: Same for the test before "should be a free position when position free". That doesn't say anything about what is actually expected to happen.

Additionally to that using "should" in test descriptions has become an anti-pattern. It's too lose in the sense that "should" can imply that something  can happen but it's also okay if it doesn't. But in our tests we want to make sure that an expectation is met. The test name should reflect that.

File: lib/board.rb (hunk 3 above)
Comment (fabientownsend) [reply lib/board.rb]: I like the idea to use the to_s for logs.

I guess I could moved that into the helper as well

File: lib/board.rb (hunk 1 above)
Comment (fabientownsend) [reply lib/board.rb]: I felt like it wasn't the best place as it's only for my tests but as both computer_spec and board_spec needed it made me take this choice.

I never use helper before, I will take a look for this solution

File: lib/board.rb [hunk 4]
```diff
@@ -52,8 +53,8 @@ def win_with_rows?(mark)
     board.any? { |row| same_symbole?(row, mark) }
   end
 
-  def same_symbole?(array, mark)
-    array.all? { |row| row == mark }
+  def same_symbole?(rows, mark)
```
Comment (christophgockel) [inline lib/board.rb:56]: `s/symbole/symbol/g`

File: lib/board.rb [hunk 5]
```diff
@@ -4,6 +4,7 @@ class Board
   attr_reader :board
   attr_reader :POSITION_MIN
   attr_reader :POSITION_MAX
+  attr_accessor :counter
```
Comment (christophgockel) [inline lib/board.rb:7]: Having a public accessor for something that is very internal to a board is something we should avoid.
What's stopping me as a client of board to call

```
board = Board.new
board.set_mark("X", 4)
board.counter = 42
```

File: spec/board_helper.rb [hunk 6]
```diff
@@ -0,0 +1,45 @@
+class BoardHelper
+  def initialize(board)
+    @board = board
+  end
+
+  def string_to_board(string_board)
+    rows = string_board.split(",")
+
+    @board.board.size.times do |index_row|
```
Comment (christophgockel) [inline spec/board_helper.rb:9]: This looks rather complicated. Isn't it possible to just use the public API of your board? As in having subsequent calls to `set_mark` while you're looping over the string?

File: spec/board_spec.rb [hunk 7]
```diff
@@ -1,71 +1,81 @@
...
-
-  it "should be free position when it never been set" do
+  it "the position 0 is free" do
     position = 0
+    board_helper.string_to_board("   ,   ,   ")
     expect(board.free_positions.include?(position)).to be true
   end
 
-  it "should shouldn't get the spot after it set" do
+  it "the position 0 isn't free" do
```
Comment (christophgockel) [inline spec/board_spec.rb:16]: When just reading the description (without looking at the code) this it makes me wonder why a board does never have a free 0th position.

**Pull request merged**